modules/navigation/nav_region.cpp

```diff
@@ -230,6 +241,60 @@ void NavRegion::update_polygons() {
 	}
 
 	surface_area = _new_region_surface_area;
+	bounds = _new_bounds;
+}
+
+void NavRegion::get_iteration_update(NavRegionIteration &r_iteration) {
+	r_iteration.navigation_layers = get_navigation_layers();
+	r_iteration.enter_cost = get_enter_cost();
+	r_iteration.travel_cost = get_travel_cost();
+	r_iteration.owner_object_id = get_owner_id();
+	r_iteration.owner_type = get_type();
+
+	r_iteration.enabled = enabled;
+	r_iteration.transform = transform;
+	r_iteration.owner_use_edge_connections = use_edge_connections;
+	r_iteration.bounds = get_bounds();
+
+	r_iteration.navmesh_polygons.resize(navmesh_polygons.size());
+
+	for (uint32_t i = 0; i < navmesh_polygons.size(); i++) {
+		const gd::Polygon &from_polygon = navmesh_polygons[i];
+		gd::Polygon &to_polygon = r_iteration.navmesh_polygons[i];
+
+		to_polygon.surface_area = from_polygon.surface_area;
+		to_polygon.owner = &r_iteration;
+		to_polygon.points.resize(from_polygon.points.size());
+
+		const LocalVector<gd::Point> &from_points = from_polygon.points;
+		LocalVector<gd::Point> &to_points = to_polygon.points;
+
+		to_points.resize(from_points.size());
+
+		for (uint32_t j = 0; j < from_points.size(); j++) {
+			to_points[j].pos = from_points[j].pos;
+			to_points[j].key = from_points[j].key;
+		}
+
+		const LocalVector<gd::Edge> &from_edges = from_polygon.edges;
+		LocalVector<gd::Edge> &to_edges = to_polygon.edges;
+
+		to_edges.resize(from_edges.size());
+
+		for (uint32_t j = 0; j < from_edges.size(); j++) {
+			const LocalVector<gd::Edge::Connection> &from_connections = from_edges[j].connections;
+			LocalVector<gd::Edge::Connection> &to_connections = to_edges[j].connections;
+
+			to_connections.resize(from_connections.size());
+
+			for (uint32_t k = 0; k < from_connections.size(); k++) {
+				to_connections[k] = from_connections[k];
+			}
+		}
+	}
+
+	r_iteration.surface_area = surface_area;
+	r_iteration.owner_rid = get_self();
```

Comment:
What are all these copies for ? To get a snapshot of the current state ?

Reply:
Yes to have a copy of the relevant region properties in case the region gets modified or even deleted while threads still query the current map iteration. That was one of the main reasons why everything was single-threaded for so long as it lead to corrupted queries or crashes.